### ISSUE TITLE
isql-v: *** The statement has grown too long on lines 243534 - 244911. Goodbye!

### ISSUE THREAD

**Issue by kotheoha** [openlink/virtuoso-opensource#721]
Hi all,

I execute various insert sql commands in graph 'http://test_graph.com' of the form:
SPARQL INSERT DATA IN \<http://test_graph.com\> { \<subject\> \<predicate\> \<object\> . };

The above commands are included in an sql file one after another. Each file includes 1000 such commands. E,g, 'INSERT_1.sql' file includes the first 1000 sql commands, 'INSERT_2.sql' file includes the next 1000 sql commands and so on. 

Specifically, for the file 'INSERT_1.sql' I used isql-v as follows:  
isql-v 1111 dba dba VERBOSE=OFF INSERT_1.sql;

The problem is that during execution I meet an error message that aborts the execution and exit. That error message is:

isql-v: *** The statement has grown too long on lines 243534 - 244911.
Already 233744 bytes read without encountering a semicolon or ampersand
in the end of line! Goodbye!

How can avoid that to not abort the execution? Further, my separate sql commands are already very small, since I use ; to each one of them at the end, as I showed in ( 1 ). Of course, the ideal is to allow any command to executed without problems, so is there any way to do that in isql-v?

Best,
KT.

**Comment by kotheoha** [comment]
Correction:
SPARQL INSERT DATA IN http://test_graph.com { <subject> <predicate> <object> .  }; ( 1 )

**Comment by HughWilliams** [comment]
@kotheoha: You seem to be seeking to perform a bulk load operation thus have you considered creating RDF datasets and loading them ? - as detailed at:

    https://virtuoso.openlinksw.com/dataspace/doc/dav/wiki/Main/VirtBulkRDFLoader

With regards to the error it seems to imply there is an insert statement with `Already 233744 bytes read without encountering a semicolon or ampersand` , which is a rather log insert statement or object values. Thus is there possibly a problem with one of the .sql files you are loading which is missing the ";"  ?

**Comment by kotheoha** [comment]
No, I do now want loading...I have already loaded my RDF data with DB.DBA.TTLP_MT(..). INSERT above is only one case, I have also DELETE and UPDATE files, in the same form as INSERT. The problem occurs in all these files.

Looking the errors, I see that Virtuoso has problems with some characters within objects like $, ---, \\. Is there any way to tell virtuoso to ignore that characters and not abort my execution? I provide such an error command below, in which Virtuoso aborts execution.

SPARQL INSERT DATA IN \<http://LGD_13_15_1M.com\> { \<node2526028091\> \<http://xmlns.com/foaf/0.1/homepage> \<http://book.bestwestern.com/bestwestern/US/CO/Aurora-hotels/BEST-WESTERN-PLUS-Gateway-Inn---Suites/Hotel-Overview.do?propertyCode=06161&cm_mmc=PTI-_-local-_-feed-_-06161> . };

**Comment by HughWilliams** [comment]
@kotheoha: What is the error you get loading that SPARQL query from a .sql file with isql ? As I get the following error:

```
$ cat x.sql 
SPARQL INSERT DATA IN <http://LGD_13_15_1M.com> { <node2526028091> <http://xmlns.com/foaf/0.1/homepage> <http://book.bestwestern.com/bestwestern/US/CO/Aurora-hotels/BEST-WESTERN-PLUS-Gateway-Inn---Suites/Hotel-Overview.do?propertyCode=06161&cm_mmc=PTI-_-local-_-feed-_-06161> . };
$ isql 1111 dba dba VERBOSE=OFF x.sql;
ERROR: Missing closing brace at line 1 of Command-Line-Load x.sql:
SPARQL INSERT DATA IN <http://LGD_13_15_1M.com> { <node2526028091> <http://xmlns.com/foaf/0.1/homepage> <http://book.bestwestern.com/bestwestern/US/CO/Aurora-hotels/BEST-WESTERN-PLUS-Gateway-Inn---Suites/Hotel-Overview.do?propertyCode=06161&cm_mmc=PTI-_-local-_-feed-_-06161> . };
```

whereas it insert fine when I use the Conductor SQL Interface ...

**Comment by kotheoha** [comment]
I did not get nothing more than that:

isql-v: *** The statement has grown too long on lines 243534 - 244911.
Already 233744 bytes read without encountering a semicolon or ampersand
in the end of line! Goodbye!

Anyway, as I said Virtuoso does not recognize some special characters within objects, like "---" in the above example case and aborts the execution. So, I try to delete them and repeat the process till not having problems. 

Yet, Virtuoso should be able to accept any character or providing conversion scripts that transform a .n3 rdf triple file to acceptable Virtuoso format. The manual process I follow is not a good way, but I have no other choice, since I want each triple to get accepted; either that belongs to load, or to insert, delete, update command.

I want each triple to get accepted for my insert, delete, update triple batches, since I use Virtuoso for comparison with other RDF systems. So, I do not want to excluding triples and make unfair comparisons because Virtuoso has some bugs in managing special characters. 

Best,
KT.

**Comment by HughWilliams** [comment]
The isql tool which is a SQL interface to Virtuoso, was not written to handle such special characters in SPARQL/SPASQL queries. we can try to make it work with such characters which would require work by development to see how this can be done without breaking other features of the isql tool.

The Virtuoso SPARQL engine does support these special character as seen by the insert working via the Conductor SQL Interface. So you should consider making a direct SPARQL or SQL connection to Virtuoso and performing the SPARUL operations directly against the Virtuoso database ...

**Comment by kotheoha** [comment]
Is there a way to "making a direct SPARQL or SQL connection to Virtuoso and performing the SPARUL operations directly against the Virtuoso database" without using the Conductor?

For instance, till now, I used isql-v in terminal and then use the commands I wanted within isql-v. Is there a similar way to do what you mentioned above without using a graphical interface (like Conductor)?

If there is, it would be convenient to provide an example by e.g. deleting a triple from an existing database (graph).

Best,
KT.

**Comment by timhaynesopenlink** [comment]
Hi,

Yes, plenty of ways. Pick a client language/environment of choice:

Web-based: there is a SPARQL endpoint at /sparql e.g. `http://localhost:8889/sparql` 

C, Python  => you probably want ODBC

R => you want either ODBC or SPARQL (there is an RSPARQL library)

Java, Open/LibreOffice => look into JDBC (direct access) or Jena (wraps SPARQL)

You can read more about the available options at https://virtuoso.openlinksw.com/dataspace/doc/dav/wiki/Main/VOSClient

HTH

**Comment by kotheoha** [comment]
Ok, thank you for the info I will check them!
Best,
KT.
